src/Orleans.Runtime/Networking/SiloConnection.cs

```diff
@@ -158,20 +165,46 @@ protected override async Task RunInternal()
         {
             try
             {
-                await Task.WhenAll(ReadPreamble(), WritePreamble());
+                if (this.connectionOptions.ProtocolVersion == NetworkProtocolVersion.Version2_0)
+                {
+                    // This version of the protocol does not support symmetric preamble, so either send or receive preamble depending on
+                    // Whether or not this is an inbound or outbound connection.
+                    if (this.RemoteSiloAddress is null)
+                    {
+                        // Inbound connection
+                        await ReadPreamble();
+                    }
+                    else
+                    {
+                        // Outbound connection
+                        await WritePreamble();
+                    }
+                }
+                else
+                {
+                    // Later versions of the protocol send and receive preamble at both ends of the connection.
+                    await Task.WhenAll(ReadPreamble(), WritePreamble());
+                }
 
                 await base.RunInternal();
             }
             finally
             {
-                if (!(this.remoteSiloAddress is null)) this.connectionManager.OnConnectionTerminated(this.remoteSiloAddress, this);
+                if (!(this.RemoteSiloAddress is null)) this.connectionManager.OnConnectionTerminated(this.RemoteSiloAddress, this);
             }
 
-            Task WritePreamble() => ConnectionPreamble.Write(this.Context, Constants.SiloDirectConnectionId, this.myAddress);
+            async Task WritePreamble()
+            {
+                await ConnectionPreamble.Write(
```

Comment:
this does not need to be async just return the task

Reply:
It's a ValueTask